akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -364,21 +364,57 @@ object ByteString {
   }
 
   /**
+   * This class helps to optimise the access time for linear access such as looping all items of [[ByteStrings]].
+   *
+   * @param lastIndex the index of the last access
+   * @param arrayIndex the array index of the last access
+   * @param seen the amount of elements being seen, excluding the elements in current array
+   */
+  private final case class LinearScanHint(lastIndex: Int, arrayIndex: Int, seen: Int)
```

Comment:
What about
```scala
case class LinearScanHint(current: ByteString1, offset: Int)

var _hint = LinearScanHint(bytestrings(0), 0)

def apply(idx: Int): Byte = {
  if (_hint.offset <= idx && idx < offset + hint.current.length)
    _hint.current(idx - _hint.offset)
  else {
    var pos = 0
    var seen = 0
    while (idx >= seen + bytestrings(pos).length) {
      seen += bytestrings(pos).length
      pos += 1
    }
    _hint = LinearScanHint(bytestrings(pos), seen)
    bytestrings(pos)(idx - seen)
  }
}
```
so all consecutive lookups to the same byte string profit (e.g. when looking up the same index several times.).